opnode/l2/driver.go

```diff
@@ -0,0 +1,208 @@
+package l2
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/event"
+
+	"github.com/ethereum-optimism/optimistic-specs/opnode/eth"
+
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type EngineDriver struct {
+	Log log.Logger
+	// API bindings to execution engine
+	RPC DriverAPI
+
+	SyncRef SyncReference
+
+	// The current driving force, to shutdown before closing the engine.
+	driveSub ethereum.Subscription
+	// There may only be 1 driver at a time
+	driveLock sync.Mutex
+
+	// Locks the L1 and L2 head changes, to keep a consistent view of the engine
+	headLock sync.RWMutex
+
+	// l1Head tracks the L1 block corresponding to the l2Head
+	l1Head eth.BlockID
+
+	// l2Head tracks the head-block of the engine
+	l2Head eth.BlockID
+
+	// l2Finalized tracks the block the engine can safely regard as irreversible
+	// (a week for disputes, or maybe shorter if we see L1 finalize and take the derived L2 chain up till there)
+	l2Finalized eth.BlockID
+
+	// The L1 block we are syncing towards, may be ahead of l1Head
+	l1Target eth.BlockID
+
+	// Genesis starting point
+	Genesis Genesis
+}
+
+// L1Head returns the block-id (hash and number) of the last L1 block that was derived into the L2 block
+func (e *EngineDriver) L1Head() eth.BlockID {
+	e.headLock.RLock()
+	defer e.headLock.RUnlock()
+	return e.l1Head
+}
+
+// L2Head returns the block-id (hash and number) of the L2 chain head
+func (e *EngineDriver) L2Head() eth.BlockID {
+	e.headLock.RLock()
+	defer e.headLock.RUnlock()
+	return e.l2Head
+}
+
+func (e *EngineDriver) Head() (l1Head eth.BlockID, l2Head eth.BlockID) {
+	e.headLock.RLock()
+	defer e.headLock.RUnlock()
+	return e.l1Head, e.l2Head
+}
+
+func (e *EngineDriver) UpdateHead(l1Head eth.BlockID, l2Head eth.BlockID) {
+	e.headLock.Lock()
+	defer e.headLock.Unlock()
+	e.l1Head = l1Head
+	e.l2Head = l2Head
+}
+
+func (e *EngineDriver) RequestHeadUpdate(ctx context.Context) error {
+	e.headLock.Lock()
+	defer e.headLock.Unlock()
+	refL1, refL2, _, err := e.SyncRef.RefByL2Num(ctx, nil, &e.Genesis)
+	if err != nil {
+		return err
+	}
+	e.l1Head = refL1
+	e.l2Head = refL2
+	return nil
+}
```

Comment:
Do these two updating functions need to be public?

Reply:
`RequestHeadUpdate` is used after initialization to try start in the happy sync without the driver having to figure it out.

`UpdateHead` is used to default to the genesis if the above `RequestHeadUpdate` fails.